### PR TITLE
Ensure that BlockMask length must always exactly match the sequence length in flex_attention

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141625
* #138788

Fixes https://github.com/pytorch/pytorch/issues/141435

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -708,7 +708,8 @@ def flex_attention_autograd(
 
     with TransformGetItemToIndex():
         input_requires_grad = any(
-            t.requires_grad for t in (query, key, value, *score_mod_other_buffers)
+            isinstance(t, torch.Tensor) and t.requires_grad
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:711]: were you hitting this??

File: torch/nn/attention/flex_attention.py [hunk 2]
```diff
@@ -359,8 +363,13 @@ def from_kv_blocks(
             BLOCK_SIZE = (BLOCK_SIZE, BLOCK_SIZE)
 
         mask_mod = mask_mod if mask_mod is not None else noop_mask
+        if seq_lengths is None:
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:366]: hmmmmmm idk bout this one

File: torch/nn/attention/flex_attention.py [hunk 3]
```diff
@@ -509,13 +524,14 @@ def _adjust(self, new_q_len: int, new_kv_len: int):
...
         )
 
-    @property
-    def shape(self):
-        """Returns the shape of the mask."""
-        *batch_dims, q_length, _ = self.kv_indices.shape
-        q_length = self.kv_indices.shape[-2] * self.BLOCK_SIZE[0]
-        kv_length = self.kv_indices.shape[-1] * self.BLOCK_SIZE[1]
-        return tuple(batch_dims + [q_length, kv_length])
+    # @property
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:527]: remove

File: torch/nn/attention/flex_attention.py [hunk 4]
```diff
@@ -1237,29 +1260,31 @@ def score_mod(
...
+            raise ValueError(
+                f"block_mask was created for block_mask.shape={block_mask.shape} but got q_len={query.size(-2)} and kv_len={key.size(-2)}. "
+                "As the block mask was created for a smaller length than you're using it for, you likely need to create a new block mask."
+            )
+        elif (
+            query.size(-2) < block_mask_q_len and key.size(-2) <= block_mask_kv_len
+        ) or (query.size(-2) <= block_mask_q_len and key.size(-2) < block_mask_kv_len):
+            raise ValueError(
+                f"block_mask was created for block_mask.shape={block_mask.shape} but got q_len={query.size(-2)} and kv_len={key.size(-2)}. "
+                "As the block mask was created for a larger length than you're using it for, you can either 1. create a new block mask with the correct length, or 2. 'adjust' the existing block mask to the correct length by calling block_mask._adjust(q_len, kv_len). This essentially 'crops' the block mask to the upper left corner, which does not work for all mask_mods!"
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:1283]: should we point to the func?

File: test/inductor/test_flex_attention.py [hunk 5]
```diff
@@ -3796,6 +3786,34 @@ def doc_mask_mod(b, h, q_idx, kv_idx):
...
+            )
+            return q, k, v
+
+        block_mask = create_block_mask(mask_mod, None, None, 1024, 1024)
+        flex_attention_call(*create_inputs(1024), block_mask=block_mask)
+        with self.assertRaisesRegex(ValueError, "block_mask was created for"):
+            flex_attention_call(*create_inputs(2048), block_mask=block_mask)
+
+        block_mask = create_block_mask(mask_mod, None, None, 1023, 1023)
+        with self.assertRaisesRegex(ValueError, "block_mask was created for"):
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:3814]: nit: stricter assert

File: torch/_higher_order_ops/flex_attention.py (hunk 1 above)
Comment (Chillee) [reply torch/_higher_order_ops/flex_attention.py:711]: Yes, the bias grads breaks score mods with captured dynamic shapes (like alibi).

File: torch/nn/attention/flex_attention.py (hunk 2 above)
Comment (Chillee) [reply torch/nn/attention/flex_attention.py:366]: I kind of want to preserve the previous flexibility if folks wanted it.

Comment (drisspg) [reply torch/nn/attention/flex_attention.py:366]: image

File: torch/_inductor/kernel/flex_attention.py [hunk 6]
```diff
@@ -1541,7 +1541,7 @@ def bwd_dq_block_mn(
 
     if not IS_FULL_BLOCKS:
         if CHECK_BLOCK_BOUNDARY:
-            mask_mod_output = tl.where(offs_n2[None, :] < KV_LEN, mask_mod_output, float("-inf"))
+            mask_mod_output = tl.where(offs_n2[None, :] < KV_LEN, mask_mod_output, False)
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:1544]: Also nit lets peel this off into early Stack so we can get this in and I can stop all those triton errors

File: torch/nn/attention/flex_attention.py (hunk 2 above)
Comment (Chillee) [reply torch/nn/attention/flex_attention.py:366]: @drisspg This doesn't really affect that - it just sets the seq_lengths to the "default" value of num_blocks * block_size.

File: torch/_inductor/kernel/flex_attention.py (hunk 6 above)
Comment (Chillee) [reply torch/_inductor/kernel/flex_attention.py:1544]: Hopefully I will land this PR soon haha.

File: test/inductor/test_flex_attention.py (hunk 5 above)
Comment (drisspg) [reply test/inductor/test_flex_attention.py:3817]: In one case the inputs are smaller than the block mask size, in the other they are bigger and we have two different error messages, just meant ensure the correct message is being shown where the asserts are the same here

File: torch/_higher_order_ops/flex_attention.py (hunk 1 above)
Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:711]: score_mod_other_buffers contains sympy expressions in that case?

File: torch/nn/attention/flex_attention.py (hunk 2 above)
Comment (drisspg) [reply torch/nn/attention/flex_attention.py:366]: Spoke offline, I disagree but wont die on this hill since this PR gets us meaningfully in a better state

**Review by drisspg — APPROVED** [review]
🍻 

Probs wait for diff train unblocking before landing

File: torch/_higher_order_ops/flex_attention.py (hunk 1 above)
Comment (Chillee) [reply torch/_higher_order_ops/flex_attention.py:711]: @drisspg yeah

**Comment by albanD** [comment]
@pytorchbot revert -m "Broken main" -c nosignal

See https://hud.pytorch.org/pytorch/pytorch/commit/795f28ac552eb61d02ea02fd64637ba814133bd8 for failures.
